File: Decorator_Pattern/go/Decorator_Pattern.go
package Decorator_Pattern

import (
	"fmt"
)

type Shape interface {
	draw()
}

type Cricle struct{}

func (*Cricle) draw() {
	fmt.Println("cricle draw")
}

type CricleDecorator interface {
	draw()
}

type RedCricle struct {
	shape Shape
}

func (s *RedCricle) draw() {
	fmt.Print("draw in decorator ")
	s.shape.draw()
}

File: Decorator_Pattern/cpp/Decorator_Pattern.cc
#include <iostream>
using namespace std;

class Shape {
public:
    virtual void draw() = 0;
};

class Cricle : public Shape {
public:
    void draw() {
        cout << "cricle draw" << endl;
    }
};

class CricleDecorator {
public:
    CricleDecorator(Shape* s):shape(s) {}
public:
    void draw() {
        cout << "draw in decorator ";
        shape->draw();
    }
private:
    Shape* shape;
};

void test_decorator() {
    Cricle c;
    c.draw();
    CricleDecorator cd(&c);
    cd.draw();
}
int main() {
    test_decorator();
    return 0;
}